### DISCUSSION TITLE
TPC-C benchmark returns 'Error in Virtual User 1: Error: key "slev" not known in dictionary'

### DISCUSSION THREAD

**Discussion by jhfoskfd** [TPC-Council/HammerDB#685] [General]
I'm brand spankin new to HammerDB so if I say something stupid please take pity and help me learn.
I'm running HammerDB v4.10 from a Windows box connecting to an Oracle 19c database on Solaris.  After building the schema and running the schema check successfully I ran a simply benchmark with 10 Vuser and all default settings except for "Total Transanctions per User" changed from 10m to 5m.  Sometimes it runs successfully to completion reporting no errors and other times it fails.  Error messages at various times have been one of the following:
- Error in Virtual User 1: Error: key "slev" not known in dictionary
- Error in Virtual User 1: Error: key "delivery" not known in dictionary

If anyone has suggestions on what might be causing these intermittent errors and how to fix it I'd appreciate the help.

**Comment by sm-shaw** [comment]
Can you provide some more details on the configuration you are using.  For example are you using the advanced XML connect pool option to configure multiple connections?  slev and delivery are 2 of the stored procedures, and this are stored in a dictionary for the XML option but not for the regular timed test, its therefore possible that a connection is lost during the test.  Also "Total Transanctions per User" only needs to be set for test scenarios where you want users to log on and off multiple times during a test.

**Comment by jhfoskfd** [comment]
Shaw,
Thanks for the reply.  XML Connect Pool is not enabled.  To hopefully make it easier to I've reproduce the problem through the cli including a couple print commands to show some of the configuration.  If there are other things I can include, other print commands ???, just let me know.  I could also upload the cfg files if you tell me which ones you need.

hammerdb>print vuconf
Virtual Users = 10
User Delay(ms) = 500
Repeat Delay(ms) = 500
Iterations = 1
Show Output = 1
Log Output = 0
Unique Log Name = 0
No Log Buffer = 0
Log Timestamps = 0

hammerdb>

hammerdb>print dict
Dictionary Settings for Oracle
connection {
 system_user     = system
 system_password = XXXXXXXXXXXXXXXXXXXX
 instance        = XXXXXXXXXXXXXXXXXXXX
 rac             = 0
}
tpcc       {
 count_ware       = 10
 num_vu           = 1
 tpcc_user        = hammerdb
 tpcc_pass        = XXXXXXXXXXXXXXXXXXXX
 tpcc_def_tab     = users
 tpcc_ol_tab      = users
 tpcc_def_temp    = temp
 partition        = false
 hash_clusters    = false
 tpcc_tt_compat   = false
 total_iterations = 5000000
 raiseerror       = false
 keyandthink      = true
 checkpoint       = true
 ora_driver       = timed
 rampup           = 1
 duration         = 5
 allwarehouse     = true
 ora_timeprofile  = true
 async_scale      = true
 async_client     = 2
 async_verbose    = false
 async_delay      = 1000
 connect_pool     = false
}

hammerdb>

hammerdb>vucreate
Script loaded, Type "print script" to view
Vuser 1 created MONITOR - WAIT IDLE
Vuser 2 created - WAIT IDLE
Vuser 3 created - WAIT IDLE
Vuser 4 created - WAIT IDLE
Vuser 5 created - WAIT IDLE
Vuser 6 created - WAIT IDLE
Vuser 7 created - WAIT IDLE
Vuser 8 created - WAIT IDLE
Vuser 9 created - WAIT IDLE
Vuser 10 created - WAIT IDLE
Vuser 11 created - WAIT IDLE
11 Virtual Users Created with Monitor VU

hammerdb>

hammerdb>vurun
Vuser 1:RUNNING
Vuser 1:Initializing xtprof time profiler
Vuser 2:RUNNING
Vuser 2:Initializing xtprof time profiler
Vuser 2:Started asynchronous clients:vuser2:ac1 vuser2:ac2
Vuser 3:RUNNING
Vuser 3:Initializing xtprof time profiler
Vuser 3:Started asynchronous clients:vuser3:ac1 vuser3:ac2
Vuser 4:RUNNING
Vuser 4:Initializing xtprof time profiler
Vuser 4:Started asynchronous clients:vuser4:ac1 vuser4:ac2
Vuser 5:RUNNING
Vuser 5:Initializing xtprof time profiler
Vuser 5:Started asynchronous clients:vuser5:ac1 vuser5:ac2
Vuser 6:RUNNING
Vuser 6:Initializing xtprof time profiler
Vuser 6:Started asynchronous clients:vuser6:ac1 vuser6:ac2
Vuser 7:RUNNING
Vuser 7:Initializing xtprof time profiler
Vuser 7:Started asynchronous clients:vuser7:ac1 vuser7:ac2
Vuser 3:VU 3 : Assigning WID=2 based on VU count 10, Warehouses = 10 (1 out of 1)
Vuser 2:VU 2 : Assigning WID=1 based on VU count 10, Warehouses = 10 (1 out of 1)
Vuser 8:RUNNING
Vuser 4:VU 4 : Assigning WID=3 based on VU count 10, Warehouses = 10 (1 out of 1)
Vuser 8:Initializing xtprof time profiler
Vuser 8:Started asynchronous clients:vuser8:ac1 vuser8:ac2
Vuser 1:Beginning rampup time of 1 minutes
Vuser 9:RUNNING
Vuser 5:VU 5 : Assigning WID=4 based on VU count 10, Warehouses = 10 (1 out of 1)
Vuser 9:Initializing xtprof time profiler
Vuser 9:Started asynchronous clients:vuser9:ac1 vuser9:ac2
Vuser 3:VU 3 : Assigning WID=2 based on VU count 10, Warehouses = 10 (1 out of 1)
Vuser 2:VU 2 : Assigning WID=1 based on VU count 10, Warehouses = 10 (1 out of 1)
Vuser 10:RUNNING
Vuser 6:VU 6 : Assigning WID=5 based on VU count 10, Warehouses = 10 (1 out of 1)
Vuser 10:Initializing xtprof time profiler
Vuser 10:Started asynchronous clients:vuser10:ac1 vuser10:ac2
Vuser 4:VU 4 : Assigning WID=3 based on VU count 10, Warehouses = 10 (1 out of 1)
Vuser 11:RUNNING
Vuser 7:VU 7 : Assigning WID=6 based on VU count 10, Warehouses = 10 (1 out of 1)
Vuser 11:Initializing xtprof time profiler
Vuser 11:Started asynchronous clients:vuser11:ac1 vuser11:ac2
Vuser 5:VU 5 : Assigning WID=4 based on VU count 10, Warehouses = 10 (1 out of 1)
Vuser 8:VU 8 : Assigning WID=7 based on VU count 10, Warehouses = 10 (1 out of 1)
Vuser 6:VU 6 : Assigning WID=5 based on VU count 10, Warehouses = 10 (1 out of 1)
Vuser 9:VU 9 : Assigning WID=8 based on VU count 10, Warehouses = 10 (1 out of 1)
Vuser 7:VU 7 : Assigning WID=6 based on VU count 10, Warehouses = 10 (1 out of 1)
Vuser 10:VU 10 : Assigning WID=9 based on VU count 10, Warehouses = 10 (1 out of 1)
Vuser 8:VU 8 : Assigning WID=7 based on VU count 10, Warehouses = 10 (1 out of 1)
Vuser 11:VU 11 : Assigning WID=10 based on VU count 10, Warehouses = 10 (1 out of 1)
Vuser 9:VU 9 : Assigning WID=8 based on VU count 10, Warehouses = 10 (1 out of 1)
Vuser 10:VU 10 : Assigning WID=9 based on VU count 10, Warehouses = 10 (1 out of 1)
Vuser 11:VU 11 : Assigning WID=10 based on VU count 10, Warehouses = 10 (1 out of 1)
Vuser 1:Rampup 1 minutes complete ...
Vuser 1:Rampup complete, Taking start AWR snapshot.
Vuser 1:Start Snapshot 35219 taken at 02 APR 2024 08:44 of instance fin45b40 (1) of database FIN45B40 (808989052)
Vuser 1:Timing test period of 5 in minutes
Vuser 1:1 ...,
Vuser 1:2 ...,
Vuser 1:3 ...,
Vuser 1:4 ...,
Vuser 1:5 ...,
Vuser 1:Test complete, Taking end AWR snapshot.
Vuser 1:End Snapshot 35220 taken at 02 APR 2024 08:49 of instance fin45b40 (1) of database FIN45B40 (808989052)
Vuser 1:Test complete: view report from SNAPID 35219 to 35220
Vuser 1:10 VU * 2 AC = 20 Active Sessions configured
Vuser 1:TEST RESULT : System achieved 24 NOPM from 54 Oracle TPM
Vuser 1:Gathering timing data from Active Virtual Users...
Vuser 2:All asynchronous clients complete
Vuser 2:FINISHED SUCCESS
Vuser 10:All asynchronous clients complete
Vuser 10:FINISHED SUCCESS
Vuser 6:All asynchronous clients complete
Vuser 6:FINISHED SUCCESS
Vuser 7:All asynchronous clients complete
Vuser 7:FINISHED SUCCESS
Vuser 4:All asynchronous clients complete
Vuser 4:FINISHED SUCCESS
Vuser 9:All asynchronous clients complete
Vuser 9:FINISHED SUCCESS
Vuser 11:All asynchronous clients complete
Vuser 11:FINISHED SUCCESSVuser 1:5 ...,
Vuser 1:Test complete, Taking end AWR snapshot.
Vuser 1:End Snapshot 35220 taken at 02 APR 2024 08:49 of instance fin45b40 (1) of database FIN45B40 (808989052)
Vuser 1:Test complete: view report from SNAPID 35219 to 35220
Vuser 1:10 VU * 2 AC = 20 Active Sessions configured
Vuser 1:TEST RESULT : System achieved 24 NOPM from 54 Oracle TPM
Vuser 1:Gathering timing data from Active Virtual Users...
Vuser 2:All asynchronous clients complete
Vuser 2:FINISHED SUCCESS
Vuser 10:All asynchronous clients complete
Vuser 10:FINISHED SUCCESS
Vuser 6:All asynchronous clients complete
Vuser 6:FINISHED SUCCESS
Vuser 7:All asynchronous clients complete
Vuser 7:FINISHED SUCCESS
Vuser 4:All asynchronous clients complete
Vuser 4:FINISHED SUCCESS
Vuser 9:All asynchronous clients complete
Vuser 9:FINISHED SUCCESS
Vuser 11:All asynchronous clients complete
Vuser 11:FINISHED SUCCESS

**Comment by jhfoskfd** [comment]
Sorry, I missed the last few lines in the cli output.

Vuser 11:FINISHED SUCCESS
Benchmark Run jobid=660BFD81B83103E263535393
hammerdb>Vuser 5:All asynchronous clients complete
Vuser 5:FINISHED SUCCESS
Vuser 3:All asynchronous clients complete
Vuser 3:FINISHED SUCCESS
Vuser 8:All asynchronous clients complete
Vuser 8:FINISHED SUCCESS
Vuser 1:Calculating timings...
Vuser 1:Writing timing data to C:/Users/WILLIA~1.CIV/AppData/Local/Temp/hdbxtprofile.log
Error in Virtual User 1: key "delivery" not known in dictionary
Vuser 1:FINISHED FAILED
ALL VIRTUAL USERS COMPLETE
hammerdb>

**Reply by sm-shaw** [reply]
OK, that is straightforward now. The error is coming from the calculation of the timing data where one (or more of the virtual users) has not run a delivery transaction during the test (or slev etc). This is because the asynchronous workload uses keying and thinking time - more is described here: https://www.hammerdb.com/blog/uncategorized/how-to-run-a-fixed-throughput-workloads/
So to avoid this error either configure more virtual users as shown in the blog e.g. 
```
Vuser 1:Test complete, Taking end Transaction Count.
Vuser 1:100 VU * 100 AC = 10000 Active Sessions configured
Vuser 1:TEST RESULT : System achieved 12612 NOPM from 29201 MariaDB TPM
```
Or use the workload without keying and thinking time. 
You will then run enough transactions and will not see this error for the timing data. 
also note that as described here https://www.hammerdb.com/blog/uncategorized/hammerdb-v4-8-new-features-viewing-your-benchmark-results-with-the-hammerdb-web-service/, hammerdb will generate graphs automatically for timing data e.g.
![Screenshot-at-2023-06-20-16-50-33-768x560](https://github.com/TPC-Council/HammerDB/assets/38044085/81bebb4a-1a7d-429e-bddf-cbe5853ac709)
and since you are using Oracle also has an option to view Oracle metrics https://www.hammerdb.com/docs/ch07s04.html to diagnose performance e.g.
![ch7-14](https://github.com/TPC-Council/HammerDB/assets/38044085/00e2dd63-66aa-4158-9d9c-7484195f5702)

**Comment by jhfoskfd** [comment]
Shaw,
Thank you for the detailed reply!  I'll try both options and let you know what I find.

**Comment by jhfoskfd** [comment]
I disabled Keying and Thinking Time.  Now I can run with 1 VU without error but setting VU > 1 results in the test running for a moment and then the HammerDB window closes.  Same result from GUI and CLI.  I also tried changing the TPROC-C Driver script form "Timed Driver Script" to "Test Driver Script" but still the same results.
In the Windows Even viewer I found this:

Faulting application name: tclsh86t.exe, version: 8.6.2.12, time stamp: 0x65ccabe5
Faulting module name: KERNELBASE.dll, version: 10.0.19041.3996, time stamp: 0xb756c9ff
Exception code: 0xebf00baa
Fault offset: 0x000000000002cf19
Faulting process id: 0x82e4
Faulting application start time: 0x01da852754f1f25a
Faulting application path: C:\Oracle_19.20.0_64Bit\HammerDB\HammerDB-4.10-Win\HammerDB-4.10\bin\tclsh86t.exe
Faulting module path: C:\WINDOWS\System32\KERNELBASE.dll
Report Id: a6f18f47-96ca-4f8a-a1f4-4097a90501f8
Faulting package full name: 
Faulting package-relative application ID:

**Comment by jhfoskfd** [comment]
Disregard the error reported in my last post.  I added a PAUSE in hammerdbcli.bat and was then able to see that it did return an error "KPEDBG_HDL_POP_FCPTRORA-24550: signal received: Unhandled exception: Code=ebf00baa Flags=0".  Although it presented differently than Oracle bug 12733000, which you mention in the documentation, and although Oracle says the underlying issue was resolved in the patch release we're running, the same fix you mentioned for bug 12733000 resolved this issue.
I still need to test the other option you mentioned above, configure more virtual users.  More to come after additional testing.
